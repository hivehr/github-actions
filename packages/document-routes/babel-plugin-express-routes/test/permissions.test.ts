import { transformFileSync } from "@babel/core";
import fs from "fs";
import { join, resolve } from "path";
import rimraf from "rimraf";

const TESTS_DIR = __dirname;
const BASE_DIR = join(TESTS_DIR, "..");
const FIXTURES_DIR = join(
    TESTS_DIR,
    "fixtures",
    "packages",
    "service",
    "src",
    "routes"
);
const OUT_DIR = join(TESTS_DIR, "routes");

const transform = (filePath: string) =>
    transformFileSync(filePath, {
        presets: ["@babel/typescript"],
        plugins: [
            [
                BASE_DIR,
                {
                    outPath: OUT_DIR,
                    maxDepth: 10
                }
            ]
        ]
    });

const verifyFile = (fileName: string) => {
    const filePath = join(FIXTURES_DIR, `${fileName}.ts`);
    transform(filePath);

    const resultFile = join(OUT_DIR, `service_${fileName}.json`);
    expect(fs.readFileSync(resultFile).toString()).toMatchSnapshot();
};

describe("Permissions", () => {
    afterAll(() => {
        rimraf.sync(OUT_DIR);
    });

    it("requireScope with single arg", () =>
        verifyFile("require_scope_with_single_arg"));

    it("requireScope with array arg", () =>
        verifyFile("require_scope_with_array_arg"));

    it("requireScope with single arg and 'or' with single arg", () =>
        verifyFile("require_scope_with_single_arg_and_or_with_single_arg"));

    it("requireScope with single arg and 'or' with array arg", () =>
        verifyFile("require_scope_with_single_arg_and_or_with_array_arg"));

    it("requireScope with array arg and 'or' with array arg", () =>
        verifyFile("require_scope_with_array_arg_and_or_with_array_arg"));

    it("route with no permissions required", () =>
        verifyFile("no_permissions_required"));

    it("require scope with single rag and two ors with single arg", () =>
        verifyFile(
            "require_scope_with_single_rag_and_two_ors_with_single_arg"
        ));

    it("test all request methods", () =>
        verifyFile("test_all_request_methods"));
});
