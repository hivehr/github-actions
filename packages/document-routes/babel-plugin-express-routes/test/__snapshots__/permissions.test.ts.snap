// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Permissions no permissions with second middleware 1`] = `
"{
  \\"serviceName\\": \\"service\\",
  \\"controllerName\\": \\"no_permissions_second_middleware\\",
  \\"methods\\": {
    \\"get\\": {
      \\"/route\\": {
        \\"permissions\\": \\"(No permissions required)\\"
      }
    }
  }
}"
`;

exports[`Permissions require scope with single rag and two ors with single arg 1`] = `
"{
  \\"serviceName\\": \\"service\\",
  \\"controllerName\\": \\"require_scope_with_single_rag_and_two_ors_with_single_arg\\",
  \\"methods\\": {
    \\"get\\": {
      \\"/route\\": {
        \\"permissions\\": \\"(over here 2) + (here) + (permission)\\"
      }
    }
  }
}"
`;

exports[`Permissions requireScope with array arg 1`] = `
"{
  \\"serviceName\\": \\"service\\",
  \\"controllerName\\": \\"require_scope_with_array_arg\\",
  \\"methods\\": {
    \\"get\\": {
      \\"/route\\": {
        \\"permissions\\": \\"(permission && permission2)\\"
      }
    }
  }
}"
`;

exports[`Permissions requireScope with array arg and 'or' with array arg 1`] = `
"{
  \\"serviceName\\": \\"service\\",
  \\"controllerName\\": \\"require_scope_with_array_arg_and_or_with_array_arg\\",
  \\"methods\\": {
    \\"get\\": {
      \\"/route\\": {
        \\"permissions\\": \\"(here && here2) + (permission && permissions1)\\"
      }
    }
  }
}"
`;

exports[`Permissions requireScope with single arg 1`] = `
"{
  \\"serviceName\\": \\"service\\",
  \\"controllerName\\": \\"require_scope_with_single_arg\\",
  \\"methods\\": {
    \\"get\\": {
      \\"/route\\": {
        \\"permissions\\": \\"(permission)\\"
      }
    }
  }
}"
`;

exports[`Permissions requireScope with single arg and 'or' with array arg 1`] = `
"{
  \\"serviceName\\": \\"service\\",
  \\"controllerName\\": \\"require_scope_with_single_arg_and_or_with_array_arg\\",
  \\"methods\\": {
    \\"get\\": {
      \\"/route\\": {
        \\"permissions\\": \\"(here && here2) + (permission)\\"
      }
    }
  }
}"
`;

exports[`Permissions requireScope with single arg and 'or' with single arg 1`] = `
"{
  \\"serviceName\\": \\"service\\",
  \\"controllerName\\": \\"require_scope_with_single_arg_and_or_with_single_arg\\",
  \\"methods\\": {
    \\"get\\": {
      \\"/route\\": {
        \\"permissions\\": \\"(here) + (permission)\\"
      }
    }
  }
}"
`;

exports[`Permissions route with no permissions required 1`] = `
"{
  \\"serviceName\\": \\"service\\",
  \\"controllerName\\": \\"no_permissions_required\\",
  \\"methods\\": {
    \\"get\\": {
      \\"/route\\": {
        \\"permissions\\": \\"(No permissions required)\\"
      }
    }
  }
}"
`;

exports[`Permissions test all request methods 1`] = `
"{
  \\"serviceName\\": \\"service\\",
  \\"controllerName\\": \\"test_all_request_methods\\",
  \\"methods\\": {
    \\"get\\": {
      \\"/route\\": {
        \\"permissions\\": \\"(permission)\\"
      }
    },
    \\"head\\": {
      \\"/route\\": {
        \\"permissions\\": \\"(permission)\\"
      }
    },
    \\"post\\": {
      \\"/route\\": {
        \\"permissions\\": \\"(permission)\\"
      }
    },
    \\"put\\": {
      \\"/route\\": {
        \\"permissions\\": \\"(permission)\\"
      }
    },
    \\"delete\\": {
      \\"/route\\": {
        \\"permissions\\": \\"(permission)\\"
      }
    },
    \\"connect\\": {
      \\"/route\\": {
        \\"permissions\\": \\"(permission)\\"
      }
    },
    \\"options\\": {
      \\"/route\\": {
        \\"permissions\\": \\"(permission)\\"
      }
    },
    \\"trace\\": {
      \\"/route\\": {
        \\"permissions\\": \\"(permission)\\"
      }
    },
    \\"patch\\": {
      \\"/route\\": {
        \\"permissions\\": \\"(permission)\\"
      }
    }
  }
}"
`;
